### PR TITLE
feat: support uint8 and int8 allreduce in tensorflow

### PR DESCRIPTION
**Pull request by kvignesh1420**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

This PR adds support for `int8` and `uint8` allreduce in tensorflow along with the relevant tests. 
Corresponding issue:  https://github.com/horovod/horovod/issues/3642

Signed-off-by: Vignesh Kothapalli <k.vignesh1420@gmail.com>

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: test/parallel/test_tensorflow.py [hunk 1]
```diff
@@ -117,19 +117,22 @@ def test_horovod_allreduce_cpu(self):
...
         size = hvd.size()
-        dtypes = self.filter_supported_types([tf.int32, tf.int64, tf.float16, tf.float32, tf.float64])
+        dtypes = self.filter_supported_types([tf.uint8, tf.int8, tf.int32, tf.int64, tf.float16, tf.float32, tf.float64])
         dims = [1, 2, 3]
         for dtype, dim in itertools.product(dtypes, dims):
             with tf.device("/cpu:0"):
-                tensor = self.random_uniform(
-                    [17] * dim, -100, 100, dtype=dtype)
+                # prevent underflows/overflows in uint8, int8
+                tensor = self.random_uniform([17] * dim, 0, 10)
```
Comment (maxhgerlach) [inline test/parallel/test_tensorflow.py:125]: I'd suggest to leave range at [-100,100] in general and add a special case for `uint8` if necessary (like taking the absolute value and casting). 

Removing any negative tensor values from the test coverage does not sound wise and I'd also prefer not to change the range with this PR where it's not necessary.

(comment applies to all tests adapted for `(u)int8`)

**Review by maxhgerlach — COMMENTED** [review]
Thanks for taking a stab at this, @kvignesh1420!

I've triggered the CI pipeline and have left one comment regarding the test script.

File: test/parallel/test_tensorflow.py (hunk 1 above)
Comment (kvignesh1420) [reply test/parallel/test_tensorflow.py:125]: sure, will change it.

**Comment by kvignesh1420** [comment]
@maxhgerlach I reverted the range back to `(-100, 100)` and fixed the tests for `int8, uint8` when `average=True`. Please take a look.

**Comment by maxhgerlach** [comment]
Thank you @kvignesh1420, that looks great!

There is a bunch of additional TensorFlow allreduce tests, scattered over a couple of separate files:
- test_tensorflow_process_sets.py
- test_xla.py
- test_xla_process_sets.py

I think it would be good to extend the `(u)int8` test coverage at least to `test_xla.py` as that one uses a different backend than `test_tensorflow.py`. 

There would be less to gain from the various tests with multiple process sets in the two other files, so I'm not sure if those are worth the busy work.

**Review by maxhgerlach — APPROVED** [review]
LGTM, let's just wait for the tests to pass. Thank you for the contribution!

**Comment by maxhgerlach** [comment]
Tests are all good. The failure at "CI / Build docker image horovod-ray" is unrelated.

**Pull request merged**